assets/src/hooks/useNearestIntersection.ts

```diff
@@ -1,13 +1,60 @@
-import { useEffect, useState } from "react"
+import { useEffect, useRef, useState } from "react"
 import { fetchNearestIntersection } from "../api"
+import {
+  LoadingResult,
+  isOk,
+  isLoadingState,
+  Err,
+  Ok,
+  Loading,
+} from "../helpers/loadingResult"
 
-export const useNearestIntersection = (
+export const useNearestIntersectionResult = (
   latitude: number,
   longitude: number
-): string | null => {
-  const [result, setResult] = useState<string | null>(null)
+): LoadingResult<string> => {
+  const latLong = { latitude, longitude }
+  const lastLatLong = useRef(latLong)
+  const [loadingState, setLoadingState] = useState<LoadingResult<string>>(
+    Loading()
+  )
+
   useEffect(() => {
-    fetchNearestIntersection(latitude, longitude).then(setResult)
+    setLoadingState(Loading())
+    ;(async () => {
+      const intersectionResult = await fetchNearestIntersection(
+        latitude,
+        longitude
+      )
+      if (intersectionResult !== null) {
+        setLoadingState(Ok(intersectionResult))
+      } else {
+        setLoadingState(
+          Err(new Error("`fetchNearestIntersection` returned `null`"))
+        )
+      }
+    })()
   }, [latitude, longitude])
-  return result
+
+  if (
+    lastLatLong.current.latitude !== latLong.latitude ||
+    lastLatLong.current.longitude !== latLong.longitude
+  ) {
+    lastLatLong.current = latLong
+    setLoadingState(Loading()) // redundant but needed to keep state synced?...
+    return Loading()
```

Comment:
It does seem to me that this is redundant and that `lastLatLong`doesn't need to be tracked. By having `[latitude, longitude]` in the dependency array of the `useEffect` hook, that hook is going to be called any time either latitude or longitude changes, so this seems to duplicate the call to `setLoadingState(Loading())` on line 23.

Reply:
The biggest issue is that when `useNearestIntersectionResult`'s input changes, it is required to invalidate the last result it got and return the `loading` state instead.

The `useState` here https://github.com/mbta/skate/pull/1873/files/a4de71efb0a7d81b53450660f8b137ae638bd5e3..1342a75301f478e8c70a8b0dc0fe5b22dc4f6286#diff-26b206fffc63b957a949bb9050dc9b9931d22ca0a6607bd710662c5f71ecdd83R18 will return the last successful value if not checked that the input has changed, which means the return will return the last loading state value, which when a new vehicle is selected, will be the last vehicles request status/result.

`useEffect` doesn't seem to run synchronously, as well as `setLoadingState` won't update the variable until next render.

Reply:
I'm struggling with the boundaries between `useNearestIntersection` and where it is ultimately used in  `VehicleNearestIntersection`. One possible proposal for an alternate way to handle invalidating the intersection when a new vehicle is selected - what if `VehicleNearestIntersection` has a key for the vehicleId? 

Check out this blog post on forcing a reset on key change [here](https://medium.com/@albertogasparin/forcing-state-reset-on-a-react-component-by-using-the-key-prop-14b36cd7448e) - I haven't personally tried that approach before outside of lists, but I'm thinking it could maybe help us to manage less state if we don't have to account for the case where the vehicle changes. This seems like it could be relevant to @lemald's thread [here](https://github.com/mbta/skate/pull/1873#discussion_r1064738877)

Reply:
THIS IS SO MUCH EASIER LOL THANK YOU! 🙇🏻‍♀️

Reply:
Interesting, I hadn't seen that approach before but it's neat!

Reply:
I had something working on the UI side but I wasn't happy with the tests, I was hitting issues similar to this https://kentcdodds.com/blog/fix-the-not-wrapped-in-act-warning where I couldn't get the promise that returns from `fetchNearestIntersection` to resolve before the test completed, which meant that the promise was firing after expect had finished for whatever reason. 

I'm instead going to walk back these changes and add a few more tests and then we should be able to merge? 
mainly trying to prioritize unblocking people by finishing this PR.

I've still got them, so we could do this in a separate PR, unsure if I should open a draft PR with the changes or just push them up as a branch.

Reply:
Knowing I can use `key` to force reset component state (e.g. to make sure it resets with it's parent component) is very helpful though!!!